services/brig/src/Brig/User/API/Auth.hs

```diff
@@ -204,25 +223,33 @@ rmCookies (uid ::: req) = do
     Auth.revokeAccess uid pw ids lls !>> authError
     return empty
 
-renew :: JSON ::: Maybe ZAuth.UserToken ::: Maybe ZAuth.AccessToken -> Handler Response
+renew :: JSON ::: Maybe (Either ZAuth.UserToken ZAuth.LegalHoldUserToken) ::: Maybe (Either ZAuth.AccessToken ZAuth.LegalHoldAccessToken) -> Handler Response
 renew (_ ::: Nothing :::  _) = throwStd authMissingCookie
-renew (_ ::: Just ut ::: at) = do
-    a <- Auth.renewAccess ut at !>> zauthError
-    tokenResponse a
+renew (_ ::: Just userToken ::: accessToken) = do
+    case (userToken, accessToken) of
+         (Left ut, Just (Left at)) -> (Auth.renewAccess ut (Just at) !>> zauthError) >>= tokenResponse
+         (Left ut, Nothing) -> Auth.renewAccess @ZAuth.User @ZAuth.Access ut Nothing !>> zauthError >>= tokenResponse
+         (Right lut, Just (Right lat)) -> Auth.renewAccess @ZAuth.LegalHoldUser @ZAuth.LegalHoldAccess lut (Just lat) !>> zauthError >>= tokenResponse
+         (Right lut, Nothing) -> Auth.renewAccess @ZAuth.LegalHoldUser @ZAuth.LegalHoldAccess lut Nothing !>> zauthError >>= tokenResponse
+         (_, _) -> throwStd authTokenMismatch
```

Comment:
What happens if we change `TokenPair u a` to `TokenPair u u` everywhere?  I suppose if that worked, it would have been done, but at least here it would nicely collapse 4 cases into two, because we wouldn't have to dynamically pattern-match into the second part of the tuple to find the type.

(I'm not sure that makes sense.  Should I try myself?)

Reply:
But TokenPair u u doesn't make much sense - The acceptable inputs are (User, Access) and (LegalHoldUser, LegalHoldAccess), but they are two different types, not twice the same.